src/bp/core/services/action/action-service.ts

```diff
@@ -54,131 +90,299 @@ export default class ActionService {
+        // bp is created here because it cannot be created in the Local Action Server thread
+        return await this._runInVm(code, dirPath, { bp: await createForAction(), ...args }, _require)
+      }
+      case 'http': {
+        return await this._runInVm(code, dirPath, { token: props.token, ...args }, _require)
+      }
+      default: {
+        throw `Unexpected runType: ${runType}`
+      }
+    }
   }
 
-  async getActionDetails(actionName: string) {
-    const action = await this.findAction(actionName)
-    const code = await this.getActionScript(action)
+  private async _runInVm(code: string, dirPath: string, args: any, _require: Function) {
+    const modRequire = new Proxy(
+      {},
+      {
+        get: (_obj, prop) => _require(prop)
+      }
+    )
 
-    const botFolder = action.location === 'global' ? 'global' : 'bots/' + this.botId
+    const vm = new NodeVM({
+      wrapper: 'none',
+      sandbox: args,
+      require: {
+        external: true,
+        mock: modRequire
+      },
+      timeout: 5000
+    })
+
+    const runner = new VmRunner()
+    return runner.runInVm(vm, code, dirPath)
+  }
+
+  private async _getActionDetails(actionName: string) {
+    const action = await this._findAction(actionName)
+    const code = await this._getActionScript(action)
+
+    const botFolder = action.scope === 'global' ? 'global' : 'bots/' + this.botId
     const dirPath = path.resolve(path.join(process.PROJECT_LOCATION, `/data/${botFolder}/actions/${actionName}.js`))
     const lookups = getBaseLookupPaths(dirPath)
 
     return { code, dirPath, lookups, action }
   }
 
+  public async loadLocalAction(actionName: string) {
+    if (yn(process.core_env.BP_EXPERIMENTAL_REQUIRE_BPFS)) {
+      await this._checkActionRequires(actionName)
+    }
+
+    const { code, dirPath, lookups } = await this._getActionDetails(actionName)
+
+    const _require = prepareRequire(dirPath, lookups)
+
+    return { code, _require, dirPath }
+  }
+
+  private _listenForCacheInvalidation() {
+    const clearDebounce = _.debounce(this._clearCache.bind(this), DEBOUNCE_DELAY, { leading: true, trailing: false })
+
+    this.cache.events.on('invalidation', key => {
+      if (key.toLowerCase().indexOf(`/actions`) > -1) {
+        clearDebounce()
+      }
+    })
+  }
+
+  private _clearCache() {
+    this._scriptsCache.clear()
+    this._globalActionsCache = undefined
+    this._localActionsCache = undefined
+    this._validScripts = {}
+    this._botsWorkspaceIdsCache.clear()
+  }
+
+  private async _getActionDefinition(file: string, scope: ActionScope): Promise<LocalActionDefinition> {
+    const name = file.replace(/\.js|\.http\.js$/i, '')
+    const legacy = !file.includes('.http.js')
+    const script = await this._getActionScript({ name, legacy, scope })
+
+    const { category, description, author, params, title, hidden } = extractMetadata(script)
+
+    return { name, scope, legacy, category, description, author, params, title, hidden }
```

Comment:
```JS
private async _getActionDefinition(file: string, scope: ActionScope): Promise<LocalActionDefinition> {
    const name = file.replace(/\.js|\.http\.js$/i, '')
    const legacy = !file.includes('.http.js')
    const script = await this._getActionScript({ name, legacy, scope })

    return { name, scope, legacy, ...extractMetadata(script)}
  }
```

Reply:
Done in 0ece0cf95f75aa523c3f98b33f396a23f719debf